File: app/lib/models/google_health_self_check.dart
class GoogleHealthSelfCheck {
  const GoogleHealthSelfCheck({
    required this.provider,
    required this.targetDate,
    required this.diagnosticsStatus,
    required this.sourceStatus,
    required this.safeToUseSleepSummary,
    required this.realHttpAttempted,
    this.session,
    required this.message,
    this.error,
  });

  final String provider;
  final String targetDate;
  final String diagnosticsStatus;
  final String sourceStatus;
  final bool safeToUseSleepSummary;
  final bool realHttpAttempted;
  final GoogleHealthSelfCheckSession? session;
  final String message;
  final String? error;

  factory GoogleHealthSelfCheck.fromJson(Map<String, dynamic> json) {
    return GoogleHealthSelfCheck(
      provider: _stringFromJson(json['provider'], fallback: 'unknown'),
      targetDate: _stringFromJson(json['target_date']),
      diagnosticsStatus: _stringFromJson(
        json['diagnostics_status'],
        fallback: 'unknown',
      ),
      sourceStatus: _stringFromJson(
        json['source_status'],
        fallback: 'unknown',
      ),
      safeToUseSleepSummary: _boolFromJson(json['safe_to_use_sleep_summary']),
      realHttpAttempted: _boolFromJson(json['real_http_attempted']),
      session: json['session'] == null
          ? null
          : GoogleHealthSelfCheckSession.fromJson(_mapFromJson(json['session'])),
      message: _stringFromJson(json['message']),
      error: _nullableStringFromJson(json['error']),
    );
  }

  String get displaySourceStatus {
    switch (sourceStatus) {
      case 'ready':
        return 'ready';
      case 'blocked':
        return 'guard blocked';
      case 'mock':
        return 'mock';
      case 'missing_auth':
        return 'auth不足';
      case 'missing_credentials':
        return 'credentials不足';
      case 'skipped':
        return 'skipped';
      case 'error':
        return 'error';
      default:
        return sourceStatus;
    }
  }

  String get displayReason {
    final sessionError = session?.error;
    if (sessionError != null && sessionError.isNotEmpty) {
      return sessionError;
    }
    if (error != null && error!.isNotEmpty) {
      return error!;
    }
    if (message.isNotEmpty) {
      return message;
    }
    return 'Google Health self-check の結果を確認してください。';
  }

  String get displayNextAction {
    if (sourceStatus == 'skipped') {
      return 'SLEEP_PROVIDER=google_health に切り替えると、sleep source 経路を確認できます。';
    }

    if (safeToUseSleepSummary && !realHttpAttempted) {
      return 'sleep summary 側の安全確認はOKです。実HTTPはまだブロックされています。';
    }

    if (safeToUseSleepSummary && realHttpAttempted) {
      return '実HTTPを含む sleep summary 経路を確認しました。';
    }

    final refresh = session?.refresh;
    if (refresh != null && refresh.error != null && refresh.error!.isNotEmpty) {
      return refresh.error!;
    }

    final api = session?.api;
    if (api != null && api.error != null && api.error!.isNotEmpty) {
      return api.error!;
    }

    final sessionError = session?.error;
    if (sessionError != null && sessionError.isNotEmpty) {
      return sessionError;
    }

    if (error != null && error!.isNotEmpty) {
      return error!;
    }

    return message.isNotEmpty ? message : 'self-check の結果を確認してください。';
  }

  static Map<String, dynamic> _mapFromJson(Object? value) {
    if (value is Map<String, dynamic>) {
      return value;
    }
    if (value is Map) {
      return value.map((key, item) => MapEntry(key.toString(), item));
    }
    return const <String, dynamic>{};
  }

  static bool _boolFromJson(Object? value, {bool fallback = false}) {
    if (value is bool) {
      return value;
    }
    if (value is num) {
      return value != 0;
    }
    if (value is String) {
      final normalized = value.trim().toLowerCase();
      if (normalized == 'true' || normalized == '1' || normalized == 'yes') {
        return true;
      }
      if (normalized == 'false' || normalized == '0' || normalized == 'no') {
        return false;
      }
    }
    return fallback;
  }

  static int? _nullableIntFromJson(Object? value) {
    if (value is int) {
      return value;
    }
    if (value is num) {
      return value.toInt();
    }
    if (value is String) {
      return int.tryParse(value.trim());
    }
    return null;
  }

  static String _stringFromJson(Object? value, {String fallback = ''}) {
    final stringValue = _nullableStringFromJson(value);
    return stringValue == null || stringValue.isEmpty ? fallback : stringValue;
  }

  static String? _nullableStringFromJson(Object? value) {
    if (value == null) {
      return null;
    }
    final text = value.toString().trim();
    return text.isEmpty ? null : text;
  }
}

class GoogleHealthSelfCheckSession {
  const GoogleHealthSelfCheckSession({
    required this.tokenAvailable,
    required this.refreshChecked,
    required this.apiRequested,
    required this.succeeded,
    this.endpoint,
    this.refresh,
    this.api,
    this.message,
    this.error,
  });

  final bool tokenAvailable;
  final bool refreshChecked;
  final bool apiRequested;
  final bool succeeded;
  final String? endpoint;
  final GoogleHealthSelfCheckRefresh? refresh;
  final GoogleHealthSelfCheckApi? api;
  final String? message;
  final String? error;

  factory GoogleHealthSelfCheckSession.fromJson(Map<String, dynamic> json) {
    return GoogleHealthSelfCheckSession(
      tokenAvailable: GoogleHealthSelfCheck._boolFromJson(
        json['token_available'],
      ),
      refreshChecked: GoogleHealthSelfCheck._boolFromJson(
        json['refresh_checked'],
      ),
      apiRequested: GoogleHealthSelfCheck._boolFromJson(json['api_requested']),
      succeeded: GoogleHealthSelfCheck._boolFromJson(json['succeeded']),
      endpoint: GoogleHealthSelfCheck._nullableStringFromJson(json['endpoint']),
      refresh: json['refresh'] == null
          ? null
          : GoogleHealthSelfCheckRefresh.fromJson(
              GoogleHealthSelfCheck._mapFromJson(json['refresh']),
            ),
      api: json['api'] == null
          ? null
          : GoogleHealthSelfCheckApi.fromJson(
              GoogleHealthSelfCheck._mapFromJson(json['api']),
            ),
      message: GoogleHealthSelfCheck._nullableStringFromJson(json['message']),
      error: GoogleHealthSelfCheck._nullableStringFromJson(json['error']),
    );
  }
}

class GoogleHealthSelfCheckRefresh {
  const GoogleHealthSelfCheckRefresh({
    required this.checked,
    required this.attempted,
    required this.requestPrepared,
    required this.realRefreshEnabled,
    required this.refreshed,
    required this.saved,
    this.message,
    this.error,
  });

  final bool checked;
  final bool attempted;
  final bool requestPrepared;
  final bool realRefreshEnabled;
  final bool refreshed;
  final bool saved;
  final String? message;
  final String? error;

  factory GoogleHealthSelfCheckRefresh.fromJson(Map<String, dynamic> json) {
    return GoogleHealthSelfCheckRefresh(
      checked: GoogleHealthSelfCheck._boolFromJson(json['checked']),
      attempted: GoogleHealthSelfCheck._boolFromJson(json['attempted']),
      requestPrepared: GoogleHealthSelfCheck._boolFromJson(
        json['request_prepared'],
      ),
      realRefreshEnabled: GoogleHealthSelfCheck._boolFromJson(
        json['real_refresh_enabled'],
      ),
      refreshed: GoogleHealthSelfCheck._boolFromJson(json['refreshed']),
      saved: GoogleHealthSelfCheck._boolFromJson(json['saved']),
      message: GoogleHealthSelfCheck._nullableStringFromJson(json['message']),
      error: GoogleHealthSelfCheck._nullableStringFromJson(json['error']),
    );
  }
}

class GoogleHealthSelfCheckApi {
  const GoogleHealthSelfCheckApi({
    required this.requested,
    required this.attempted,
    required this.requestPrepared,
    required this.realApiEnabled,
    required this.succeeded,
    this.statusCode,
    this.message,
    this.error,
  });

  final bool requested;
  final bool attempted;
  final bool requestPrepared;
  final bool realApiEnabled;
  final bool succeeded;
  final int? statusCode;
  final String? message;
  final String? error;

  factory GoogleHealthSelfCheckApi.fromJson(Map<String, dynamic> json) {
    return GoogleHealthSelfCheckApi(
      requested: GoogleHealthSelfCheck._boolFromJson(json['requested']),
      attempted: GoogleHealthSelfCheck._boolFromJson(json['attempted']),
      requestPrepared: GoogleHealthSelfCheck._boolFromJson(
        json['request_prepared'],
      ),
      realApiEnabled: GoogleHealthSelfCheck._boolFromJson(
        json['real_api_enabled'],
      ),
      succeeded: GoogleHealthSelfCheck._boolFromJson(json['succeeded']),
      statusCode: GoogleHealthSelfCheck._nullableIntFromJson(
        json['status_code'],
      ),
      message: GoogleHealthSelfCheck._nullableStringFromJson(json['message']),
      error: GoogleHealthSelfCheck._nullableStringFromJson(json['error']),
    );
  }
}

File: backend/app/services/google_health_self_check.py
from __future__ import annotations

from datetime import date

from app.config import AppConfig
from app.models.google_health import (
    GoogleHealthApiRequestPreviewModel,
    GoogleHealthProviderErrorSummaryModel,
    GoogleHealthSelfCheckApiModel,
    GoogleHealthSelfCheckRefreshModel,
    GoogleHealthSelfCheckResponse,
    GoogleHealthSelfCheckSessionModel,
)
from app.services.google_health_diagnostics import get_google_health_diagnostics
from app.services.google_health_service import GOOGLE_HEALTH_PROVIDER_NAME
from app.services.google_health_session import (
    GoogleHealthApiClientSummary,
    GoogleHealthSessionResult,
    GoogleHealthTokenRefreshSummary,
)
from app.services.google_health_sleep_source import fetch_google_health_sleep_summary
from app.services.google_health_token_store import GoogleHealthTokenStore


GOOGLE_HEALTH_SELF_CHECK_STATUS_SKIPPED = "skipped"


def run_google_health_self_check(
    *,
    config: AppConfig,
    target_date: date | None = None,
    token_store: GoogleHealthTokenStore | None = None,
) -> GoogleHealthSelfCheckResponse:
    """
    Run a safe Google Health smoke path for local development.

    The check uses the same source/session/API-client boundary as
    /sleep/summary with SLEEP_PROVIDER=google_health, but returns only safe
    status metadata. It never returns tokens, authorization headers, client
    secrets, or raw health payloads.
    """

    check_date = target_date or date.today()
    store = token_store or GoogleHealthTokenStore()
    diagnostics = get_google_health_diagnostics(config, token_store=store)

    if config.sleep_provider != GOOGLE_HEALTH_PROVIDER_NAME:
        return GoogleHealthSelfCheckResponse(
            provider=GOOGLE_HEALTH_PROVIDER_NAME,
            target_date=check_date.isoformat(),
            diagnostics_status=diagnostics.overall_status,
            source_status=GOOGLE_HEALTH_SELF_CHECK_STATUS_SKIPPED,
            safe_to_use_sleep_summary=False,
            real_http_attempted=False,
            session=None,
            message=(
                "Google Health self-check skipped because SLEEP_PROVIDER is "
                "not google_health. Set SLEEP_PROVIDER=google_health to test "
                "the sleep source path."
            ),
            error=diagnostics.error,
        )

    source_result = fetch_google_health_sleep_summary(
        config=config,
        target_date=check_date,
        token_store=store,
    )
    session_model = _build_session_model(source_result.session)
    real_http_attempted = bool(
        session_model
        and session_model.api
        and session_model.api.attempted
    )

    return GoogleHealthSelfCheckResponse(
        provider=GOOGLE_HEALTH_PROVIDER_NAME,
        target_date=check_date.isoformat(),
        diagnostics_status=diagnostics.overall_status,
        source_status=source_result.status,
        safe_to_use_sleep_summary=source_result.summary.available,
        real_http_attempted=real_http_attempted,
        session=session_model,
        message=source_result.message or source_result.summary.message or "",
        error=source_result.error,
    )


def _build_session_model(
    session: GoogleHealthSessionResult | None,
) -> GoogleHealthSelfCheckSessionModel | None:
    if session is None:
        return None

    return GoogleHealthSelfCheckSessionModel(
        token_available=session.token_available,
        refresh_checked=session.refresh_checked,
        api_requested=session.api_requested,
        succeeded=session.succeeded,
        endpoint=session.endpoint,
        refresh=_build_refresh_model(session.refresh_summary),
        api=_build_api_model(session.api_summary),
        message=session.message,
        error=session.error,
    )


def _build_refresh_model(
    refresh: GoogleHealthTokenRefreshSummary | None,
) -> GoogleHealthSelfCheckRefreshModel | None:
    if refresh is None:
        return None

    return GoogleHealthSelfCheckRefreshModel(
        checked=True,
        attempted=refresh.attempted,
        request_prepared=refresh.request_prepared,
        real_refresh_enabled=refresh.real_refresh_enabled,
        refreshed=refresh.refreshed,
        saved=refresh.saved,
        message=refresh.message,
        error=refresh.error,
    )


def _build_api_model(
    api: GoogleHealthApiClientSummary | None,
) -> GoogleHealthSelfCheckApiModel | None:
    if api is None:
        return None

    preview = api.request_preview
    preview_model = None
    if preview is not None:
        preview_model = GoogleHealthApiRequestPreviewModel(
            endpoint=preview.endpoint,
            method=preview.method,
            has_bearer_auth=preview.has_bearer_auth,
            query_param_keys=list(preview.query_param_keys),
            query_params=dict(preview.query_params),
            preview_url=preview.preview_url,
        )

    provider_error_summary_model = None
    if api.provider_error_summary is not None:
        provider_error_summary_model = GoogleHealthProviderErrorSummaryModel(
            http_status_code=api.provider_error_summary.http_status_code,
            provider_error_code=api.provider_error_summary.provider_error_code,
            provider_error_status=api.provider_error_summary.provider_error_status,
            provider_error_message_hint=api.provider_error_summary.provider_error_message_hint,
            provider_error_reason=api.provider_error_summary.provider_error_reason,
            provider_error_domain=api.provider_error_summary.provider_error_domain,
            provider_error_metadata_keys=list(
                api.provider_error_summary.provider_error_metadata_keys
            ),
            www_authenticate_hint=api.provider_error_summary.www_authenticate_hint,
            suggested_cause=api.provider_error_summary.suggested_cause,
        )

    return GoogleHealthSelfCheckApiModel(
        requested=True,
        attempted=api.attempted,
        request_prepared=api.request_prepared,
        real_api_enabled=api.real_api_enabled,
        succeeded=api.succeeded,
        status_code=api.status_code,
        request_preview=preview_model,
        provider_error_category=api.provider_error_category,
        provider_error_summary=provider_error_summary_model,
        message=api.message,
        error=api.error,
    )
